papermill/execute.py

```diff
@@ -13,26 +15,67 @@
+
+    We are doing this for the following reasons:
+    1. Notebooks will stop executing when they encounter a failure but not raise a CellException.
+       This allows us to save the notebook with the traceback even though a CellExecutionError
+       was encountered.
 
-    def preprocess(self, nb, resources):
-        """
-        The default preprocess behavior if allow_errors = True is to continue executing all the cells.
-        We want the notebook to cease execution and then write out the state of the notebook with the traceback
-        on the cell that failed.
-        """
-        # Clear out all outputs prior to execution.
-        for cell in nb.cells:
-            cell.outputs = []
+    2. We want to write the notebook as cells are executed. We inject our logic for that here.
 
-        # Catch CellExecutionError if thrown which lets this method finish so we can write the state of the notebook
-        # to disk.
-        try:
-            nb, resources = super(PapermillExecutePreprocessor, self).preprocess(nb, resources)
-        except CellExecutionError:
-            pass
+    3. We want to include timing and execution status information with the metadata of each cell.
 
-        return nb, resources
+    Parameters
+    ----------
+    nb : NotebookNode
+        Notebook being converted
+    resources : dictionary
+        Additional resources used in the conversion process.  Allows
+        preprocessors to pass variables into the Jinja engine.
+    """
+    output_path = nb.metadata.papermill['output_path']
+
+    # Reset the notebook.
+    for cell in nb.cells:
+        # Reset the cell execution counts.
+        if hasattr(cell, "execution_count"):
+            cell.execution_count = None
+
+        # Clear out the papermill metadata for each cell.
+        cell.metadata['papermill'] = dict(
+            exception=None,
+            start_time=None,
+            end_time=None,
+            duration=None
+        )
+        cell.outputs = []
+
+    # Execute each cell and update the output in real time.
+    with futures.ThreadPoolExecutor(max_workers=1) as executor:
+        for index, cell in enumerate(nb.cells):
+            cell.execution_count = "*"
+            future = executor.submit(write_ipynb, nb, output_path)
+            t0 = datetime.datetime.utcnow()
+            try:
+                nb.cells[index], resources = self.preprocess_cell(cell, resources, index)
+                cell.metadata['papermill']['exception'] = False
+            except CellExecutionError:
+                cell.metadata['papermill']['exception'] = True
+                break
+            finally:
+                future.result()
+                t1 = datetime.datetime.utcnow()
+                cell.metadata['papermill']['start_time'] = t0.isoformat()
+                cell.metadata['papermill']['end_time'] = t1.isoformat()
+                cell.metadata['papermill']['duration'] = (t1 - t0).total_seconds()
```

Comment:
You should move `future.result()` after your time calculations, otherwise you're also measuring IO. This will become pretty glaring with larger notebooks.